packetbeat/protos/dns/dns_tcp_test.go

```diff
@@ -0,0 +1,648 @@
+
+	dns.Parse(packet, tcptuple, tcp.TcpDirectionOriginal, private)
+	m := expectResult(t, dns)
+	assert.Equal(t, "tcp", mapValue(t, m, "transport"))
+	assert.Nil(t, mapValue(t, m, "bytes_in"))
+	assert.Equal(t, len(q.response), mapValue(t, m, "bytes_out"))
+	assert.Nil(t, mapValue(t, m, "responsetime"))
+	assert.Equal(t, common.ERROR_STATUS, mapValue(t, m, "status"))
+	assert.Equal(t, OrphanedResponse.Error(), mapValue(t, m, "notes"))
+	assertMapStrData(t, m, q)
+}
+
+// Verify that the first request is published without a response and that
+// the status is error. This second packet will remain in the transaction
+// map awaiting a response.
+func TestParseTcp_errorDuplicateRequests(t *testing.T) {
+	var private protos.ProtocolData
+	dns := newDns(testing.Verbose())
+	q := elasticATcp
+	packet := newPacket(forward, q.request)
+	tcptuple := testTcpTuple()
+
+	dns.Parse(packet, tcptuple, tcp.TcpDirectionOriginal, private)
+	assert.Equal(t, 1, dns.transactions.Size(), "There should be one transaction.")
+
+	dns.Parse(packet, tcptuple, tcp.TcpDirectionOriginal, private)
+	// The first request is published and this one becomes a transaction
+	assert.Equal(t, 1, dns.transactions.Size(), "There should be one transaction.")
+
+	m := expectResult(t, dns)
+	assertRequest(t, m, q)
+	assert.Equal(t, "tcp", mapValue(t, m, "transport"))
+	assert.Equal(t, len(q.request), mapValue(t, m, "bytes_in"))
+	assert.Nil(t, mapValue(t, m, "bytes_out"))
+	assert.Nil(t, mapValue(t, m, "responsetime"))
+	assert.Equal(t, common.ERROR_STATUS, mapValue(t, m, "status"))
+	assert.Equal(t, DuplicateQueryMsg.Error(), mapValue(t, m, "notes"))
+}
+
+// Same than the previous one but on the same stream
+// Checks that PrepareNewMessage and Parse can manage two messages on the same stream, in different packets
+func TestParseTcp_errorDuplicateRequestsOneStream(t *testing.T) {
+	var private protos.ProtocolData
+	dns := newDns(testing.Verbose())
+	q := elasticATcp
+	packet := newPacket(forward, q.request)
+	tcptuple := testTcpTuple()
+
+	private = dns.Parse(packet, tcptuple, tcp.TcpDirectionOriginal, private)
+	assert.Equal(t, 1, dns.transactions.Size(), "There should be one transaction.")
+
+	dns.Parse(packet, tcptuple, tcp.TcpDirectionOriginal, private)
+	// The first query is published and this one becomes a transaction
+	assert.Equal(t, 1, dns.transactions.Size(), "There should be one transaction.")
+
+	m := expectResult(t, dns)
+	assertRequest(t, m, q)
+	assert.Equal(t, "tcp", mapValue(t, m, "transport"))
+	assert.Equal(t, len(q.request), mapValue(t, m, "bytes_in"))
+	assert.Nil(t, mapValue(t, m, "bytes_out"))
+	assert.Nil(t, mapValue(t, m, "responsetime"))
+	assert.Equal(t, common.ERROR_STATUS, mapValue(t, m, "status"))
+	assert.Equal(t, DuplicateQueryMsg.Error(), mapValue(t, m, "notes"))
+}
+
+// Checks that PrepareNewMessage and Parse can manage two messages sharing one packet on the same stream
+// It typically happens when a SOA is followed by AXFR
```

Comment:
Right, the comment is misleading. It would be better with two requests with unique ID.
But still, I think the test does the job with a short amount of lines of code. The test shows that two requests that share one connection (and one packet during the connection) will get parsed and create  a transaction.